### DISCUSSION TITLE
Catapult design question

### DISCUSSION THREAD

**Discussion by nthuyouwei** [bol-edu/HLS-SOC-Discussions#223] [Q&A]
# 老師，助教們好，我目前有個design schedule如下圖:

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145022311/7c026a89-0f96-4f77-99b5-044e05345fab)
# 目前我們已經實現在vitis_hls,對應code如下圖:

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145022311/a374d878-540a-4c0a-a824-550dab0d7985)
(https://github.com/vic9112/PQC_Falcon/blob/main/optimized/combined/fiFFNTT_ver9/hls_code/fiFFNTT.cpp)

# 我們可以利用#pragma限制指合成一塊PE，但我今天要把code轉成Catapult hls_code時類似coding方法(如下圖):

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145022311/4fe51a6a-1599-4fff-ae5e-7fb9ff65cb0c)

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145022311/ba48c574-f1ac-4fd5-a617-bc3806af98fd)

# 就會出現以下這個error(但如果用inline就會不會有error，但這並是我想要的)，不知如何解決。

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145022311/a5054a71-bfe9-4d3f-897a-7ef1fc29cd68)
#  我後面把PE寫成Class 把他寫到top design利用一個參數當作ac_channel 連起來一樣會跑出相同的error。

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145022311/180655f7-d794-4a31-ae26-9f1e2d4b9807)

目前design 放置在 https://github.com/vic9112/PQC_Falcon/tree/main/Catapult ，不知道熟悉Catapult的老師或助教們有沒有更好的coding方式可以提供參考。

**Comment by SiemensCatapultAE** [comment]
同學你好，

這問題在於memory architecture，假設這design是fully pipeline的架構，也就是各個PE是並行的，各PE間大量的資料傳送應使用ping-pong memory architecture，Catapult支持二種ping-pong memory寫法，我把slide和lab附上，若有需要，我們可以開個線上會議討論下。
[03_Design_With_Shared_Memory_C++.pdf](https://github.com/bol-edu/HLS-SOC-Discussions/files/15426341/03_Design_With_Shared_Memory_C%2B%2B.pdf)
[03_Shared_Mem.zip](https://github.com/bol-edu/HLS-SOC-Discussions/files/15426396/03_Shared_Mem.zip)

**Comment by nthuyouwei** [comment]
老師好，不知道到您何時有空，希望可以跟您約個半小時的時間討論一下。因為上面的方法好像是要建立兩個PE之間的interconnect，我們目前是只希望有**一個**PE而他可以重複設定參數在跑第二次。同時PE要去externel SRAM取資料跟寫資料(並不會照順序)，所以我們並不太確定要如何讓他用同一個PE連續跑兩次以上。

**Comment by SiemensCatapultAE** [comment]
同學你好,

我們約下周一上午10點 (5/27)討論，是否ok ?

**Comment by nthuyouwei** [comment]
可以的！到時連結您在公布於討論區或是可以寄件於我信箱 ：
gungod0487@gmail.com

**Reply by SiemensCatapultAE** [reply]
已寄送鏈結，https://teams.microsoft.com/l/meetup-join/19%3ameeting_MjA2ZDA0OWYtZjRjNy00NDVkLThhNTItMDE5NWNmYzdiZjUx%40thread.v2/0?context=%7b%22Tid%22%3a%2238ae3bcd-9579-4fd4-adda-b42e1495d55a%22%2c%22Oid%22%3a%22592003fa-644d-40a5-857c-93a9506eb20c%22%7d

**Comment by boledulab** [comment]
From my understanding of your design intention is to use one PE with separate input RAM and output RAM. The RAM are alternatively used as input and output for successive PE stage.  A possible code structure is
1. declare two array  RAM1, and RAM2
2.  for( i = 0; i < 10 ; i++) {
        odd = i%2;     // odd = 1; RAM1 as input, RAM2 as output
        PE( RAM1, RAM2, odd ...  other input arguments  )
    }
    // note: the loop is rolled, so there is one PE instantiated
3. PE (  )
       for(    )  {
         if(odd )
            data = RAM1(i);   // data from RAM1
         else
            data = RAM2(i);  // data from RAM2
         processing data;   generate data_out;
       
       if(odd )
          RAM2(i)  = data_out;
      else
          RAM1(i)  = data_out;

This coding style can be applied to both Vitis and Catapult, though Catapult may use class method.
